File: 2020/dec5/part1.exs
IO.puts("Hello, December 5th")

defmodule Seating do
    def getID(input) do
        {r, c} = Enum.split(String.graphemes(input), 7)
        row = Enum.reduce(r, %{:s=>0, :e=>127}, &Seating.reducer/2)
        col = Enum.reduce(c, %{:s=>0, :e=>7}, &Seating.reducer/2)
        row.s*8+col.s
    end

    def reducer("F", %{:s=>s, :e=>e}), do: %{s: s, e: floor(s + (e-s)/2)}
    def reducer("L", %{:s=>s, :e=>e}), do: %{s: s, e: floor(s + (e-s)/2)}
    def reducer("B", %{:s=>s, :e=>e}), do: %{s: floor(s + (e-s)/2)+1, e: e}
    def reducer("R", %{:s=>s, :e=>e}), do: %{s: floor(s + (e-s)/2)+1, e: e}
end

stream = File.stream!("input.txt")
lines = Enum.map(stream, &String.trim/1)
IO.puts(Enum.max(Enum.map(lines, &Seating.getID/1)))

File: 2020/dec18/part1.swift
import Swift

print("Hello, December 18th")

enum ops {
    case none
    case add
    case mul
}

class reader {
    var input = ""
    var curr: Character = "\u{00}"

    init(input i: String) {
        input = i
    }

    func next() -> Bool {
        if input.count  == 0 {
            return false
        }
        curr = input[input.startIndex]
        input = String(input.suffix(input.count - 1))
        return true
    }
}

class maffs {
    var a = 0 // accumulator, treats the first op as an add always
    var op = ops.add

    var v = ""
    var stream:reader?

    init(reader r:reader) {
        stream = r
    }

    func quick() -> Int {
        if stream == nil {
            return 0
        }
        while stream!.next(){
            if stream!.curr == " " {
                interpret()
                continue
            }else if stream!.curr == "(" {
                v = String(maffs(reader: stream!).quick())
                continue
            }else if stream!.curr == ")" {
                interpret()
                return a
            }
            v.append(stream!.curr)
        }
        interpret()
        return a
    }

    func interpret() {
        if op == ops.none {
            if v == "+" {
                op = ops.add
            }else if v == "*" {
                op = ops.mul
            }
            v = ""
            return
        }
        
        let b = Int(v)
        v = ""

        if op == ops.add {
            a += b!
        }else if op == ops.mul {
            a *= b!
        }
        op = ops.none
    }
}

import Foundation

let someFile = "input.txt"
let input = try! String(contentsOfFile: someFile)
var result = 0

let lines = input.split(separator:"\n")

for line in lines {
    let m = maffs(reader: reader(input: String(line)))
    result += m.quick()
}

print(result)